### PR TITLE
Expose the initial logger creation error

### PR DESCRIPTION
**Pull request by riversand963**
#9984 changes the behavior of RocksDB: if logger creation failed during `SanitizeOptions()`,
`DB::Open()` will fail. However, since `SanitizeOptions()` is called in `DBImpl::DBImpl()`, we cannot
directly expose the error to caller without some additional work.
This is a first version proposal which:
- Adds a new member `init_logger_creation_s` to `DBImpl` to store the result of init logger creation
- Checks the error during `DB::Open()` and return it to caller if non-ok

This is not very ideal. We can alternatively move the logger creation logic out of the `SanitizeOptions()`.
Since `SanitizeOptions()` is used in other places, we need to check whether this change breaks anything
in case other callers of `SanitizeOptions()` assumes that a logger should be created.

Test plan
make check

### CODE REVIEW

File: db/db_impl/db_impl_open.cc [hunk 1]
```diff
@@ -59,6 +60,9 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src,
     if (!s.ok()) {
       // No place suitable for logging
       result.info_log = nullptr;
+      if (logger_creation_s) {
+        *logger_creation_s = s;
```
Comment (pdillinger) [inline db/db_impl/db_impl_open.cc:64]: We have ASSERT_STATUS_CHECKED failures; if the simplest solution is to PermitUncheckedError() here, that's OK with me (not ideal)

**Review by pdillinger — APPROVED** [review]
Thanks for the fix!

File: db/db_impl/db_impl_open.cc (hunk 1 above)
Comment (riversand963) [reply db/db_impl/db_impl_open.cc:64]: A single `PermitUncheckedError()` here does not fully fix the issue. We can either:
- Put a PermitUncheckedError() in ~DBImpl(), or
- For each of CompactedDB, DBImplSecondary, DBImplReadOnly, explicitly call `PermitUncheckedError()` shortly after we are done with logger creation.

The latest commit goes by the second approach.

Comment (riversand963) [reply db/db_impl/db_impl_open.cc:64]: To keep this fix simple, updated the PR to take the first approach.

**Pull request closed without merging**